src/software/ai/navigator/path_planner/global_path_planner.cpp

```diff
@@ -0,0 +1,59 @@
+#include "software/ai/navigator/path_planner/global_path_planner.h"
+
+GlobalPathPlannerFactory::GlobalPathPlannerFactory(
+    const std::shared_ptr<const RobotNavigationObstacleConfig> navigation_obstacle_config,
+    const World &world, const Rectangle &navigable_area)
+{
+    RobotNavigationObstacleFactory obstacle_factory =
+        RobotNavigationObstacleFactory(navigation_obstacle_config);
+    std::vector<MotionConstraint> all_constraints = allValuesMotionConstraint();
+
+    // The idea of this is similar to Gray codes
+    // (https://en.wikipedia.org/wiki/Gray_code#History_and_practical_application). The
+    // idea is that each bit in the counter represents a MotionConstraint and whether it
+    // is on and off. By cycling through every combination of bits, we'll consequently
+    // also have every combination of MotionConstraints
+    // (https://www.geeksforgeeks.org/generate-n-bit-gray-codes/)
+    for (unsigned counter = 0; counter < std::pow(2, all_constraints.size()); ++counter)
```

Comment:
I'm curious if we need to generate every single possible combination of motion constraints, or if there are scenarios which we can skip over? If there are only a few cases which can never happen, then it may be just easier to calculate every single combination, but if there are many scenarios which can never happen, it may be a good idea to filter some of them out to reduce run time and memory usage. 256 * 0.09m grid field seems like a lot of memory.

Reply:
Not sure if we create list of motion constraints elsewhere, but maybe we could somehow limit the set of them to what can be created in [motion_constraint_set_builder](https://github.com/UBC-Thunderbots/Software/blob/ec00976f544f79419bc86fb44e61a14704650390/src/software/ai/motion_constraint/motion_constraint_set_builder.cpp) and [motion_constraint_visitor](https://github.com/UBC-Thunderbots/Software/blob/ec00976f544f79419bc86fb44e61a14704650390/src/software/ai/motion_constraint/motion_constraint_visitor.cpp)

Reply:
iirc you mentioned that it takes ~1 sec to initialize the ENLSVG path planners. Curious if that was with these 256 combinations?

Reply:
That time is a bit off actually. I got a bit confused by the test infrastructure runtime

Reply:
Do you know roughly how long it takes? I guess with `HALF_METER_AROUND_BALL` and `AVOID_BALL_PLACEMENT_INTERFERENCE` not being considered anymore by Global Path Planner, it will be 64 total combinations now!?

Reply:
So all the constraints will take: 2100 ms in Div A, 1300 ms in Div B

Reply:
I can make exceptions for these constraints, do you think it's worth the time improvement to do this for now?

Reply:
No probably not. We should just keep it in mind for later if we deem it to slow.